### PR TITLE
feat(whiteboardStore): preload file when the whiteboard scene state has been changed

### PR DESCRIPTION
**Pull request by Cheerego7**

### CODE REVIEW

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 1]
```diff
@@ -149,6 +150,19 @@ export class WhiteboardStore {
                     if (modifyState.broadcastState) {
                         this.updateViewMode(modifyState.broadcastState.mode);
                     }
+
+                    //  preload file when the whiteboard scene state has been changed.
+                    if (room.state.sceneState.scenes[0].ppt) {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]: why is `scenes[0]`? why can't you use 1 or other numbers ?

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 2]
```diff
@@ -149,6 +150,19 @@ export class WhiteboardStore {
...
+                        const pptFiles = /(static|dynamic)Convert\/([0-9a-f]{32})\//.exec(
+                            room.state.sceneState.scenes[0].ppt.src,
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```ts
// utils.ts
const getWhiteboardPPTInfo = (pptResources?: { src: string }): null | { pptType: "static" | "dynamic", taskUUID: string } => {
    if (typeof pptResources === "undefined") {
        return null;
    }

    const resourcesInfo = /(static|dynamic)Convert\/([0-9a-f]{32})\//.exec(pptResources.src);

    if (!resourcesInfo) {
        return null;
    }

    const [, pptType, taskUUID] = resourcesInfo;
    return {
        pptType,
        taskUUID,
    };
}

// WhiteboardStore.ts
const pptResources = getWhiteboardPPTInfo(room.state.sceneState.scenes[0].ppt);

if (pptResources === null) {
    return;
}

const { pptType, taskUUID } = pptResources;

void getCoursewarePreloader().preload(taskUUID, pptType);

```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: desktop/renderer-app/src/stores/WhiteboardStore.ts (hunk 1 above)
Comment (Cheerego7) [reply desktop/renderer-app/src/stores/WhiteboardStore.ts]: scenes[0] is first ppt page

**Review by hyrious — DISMISSED** [review]

File: desktop/renderer-app/src/utils/CoursewarePreloader.ts [hunk 3]
```diff
@@ -79,10 +84,38 @@ class CoursewarePreloader {
...
     }
+
+    public deletePreloadTasks(taskUUID: TaskUUID): void {
+        if (this.preloadTasks.has(taskUUID)) {
+            window.clearTimeout(this.preloadTasks.get(taskUUID));
+            this.preloadTasks.delete(taskUUID);
+        }
+    }
+
+    public matchPptSrcToPreload(pptSrc: string): void {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/utils/CoursewarePreloader.ts]: payloadByPPTResources

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 4]
```diff
@@ -149,6 +150,13 @@ export class WhiteboardStore {
                     if (modifyState.broadcastState) {
                         this.updateViewMode(modifyState.broadcastState.mode);
                     }
+
+                    // TODO: That is temporarily plan, use RTM emit message is follow-up plan.
+                    //  preload file when the whiteboard scene state has been changed.
+                    const pptSrc = modifyState.sceneState?.scenes[0]?.ppt?.src;
+                    if (pptSrc) {
+                        void getCoursewarePreloader().matchPptSrcToPreload(pptSrc);
+                    }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]: getCoursewarePreloader().matchPptSrcToPreload(modifyState.sceneState?.scenes[0]?.ppt);

File: desktop/renderer-app/src/utils/CoursewarePreloader.ts [hunk 5]
```diff
@@ -79,10 +84,38 @@ class CoursewarePreloader {
...
+        const pptFiles = /(static|dynamic)Convert\/([0-9a-f]{32})\//.exec(pptSrc);
+        if (!pptFiles) return;
+        const [, pptType, taskUUID] = pptFiles;
+
+        this.deletePreloadTasks(taskUUID);
```
Comment (BlackHole1) [inline desktop/renderer-app/src/utils/CoursewarePreloader.ts]: why delete

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: desktop/renderer-app/src/utils/CoursewarePreloader.ts (hunk 5 above)
Comment (Cheerego7) [reply desktop/renderer-app/src/utils/CoursewarePreloader.ts]: debounce

**Review by BlackHole1 — APPROVED** [review]

**Pull request merged**